library/pod-security-policy/users/src.rego

```diff
@@ -1,19 +1,39 @@
 package k8spspallowedusers
 
 violation[{"msg": msg}] {
-  rule := input.parameters.runAsUser.rule
-  input_containers[input_container]
-  provided_user := run_as_user(input_container, input.review)
+  fields := ["runAsUser", "runAsGroup", "supplementalGroups", "fsGroup"]
+  field := fields[_]
+  container := input_containers[_]
+  msg := get_type_violation(field, container)
+}
+
+get_type_violation(field, container) = msg {
+  field == "runAsUser"
+  has_field(input.parameters, field)
+  params := input.parameters[field]
+  msg := get_user_violation(params, container)
+}
+
+get_type_violation(field, container) = msg {
+  field != "runAsUser"
+  has_field(input.parameters, field)
+  params := input.parameters[field]
+  msg := get_violation(field, params, container)
+}
+
+# RunAsUser (separate due to "MustRunAsNonRoot")
+get_user_violation(params, container) = msg {
+  rule := params.rule
+  provided_user := get_field_value("runAsUser", container, input.review)
+  not is_null(provided_user)
   not accept_users(rule, provided_user)
-  msg := sprintf("Container %v is attempting to run as disallowed user %v", [input_container.name, provided_user])
+  msg := sprintf("Container %v is attempting to run as disallowed user %v. Allowed runAsUser: %v", [container.name, provided_user, params])
 }
 
-violation[{"msg": msg}] {
-  rule := input.parameters.runAsUser.rule
-  input_containers[input_container]
-  not run_as_user(input_container, input.review)
-  rule != "RunAsAny"
-  msg := sprintf("Container %v is attempting to run without a required securityContext/runAsUser", [input_container.name])
+get_user_violation(params, container) = msg {
+  is_null(get_field_value("runAsUser", container, input.review))
```

Comment:
`is_null()` works, but seems slightly more brittle than testing for "falsey" values, as if a function inadvertently doesn't cover all possible states, it may return `undefined`

Reply:
I've re-written the rego to remove `is_null`